feature-governance-impl/src/main/java/io/novafoundation/nova/feature_governance_impl/presentation/tindergov/cards/TinderGovCardsViewModel.kt

```diff
@@ -123,14 +143,40 @@ class TinderGovCardsViewModel(
         }
     }
 
+    private fun loadFirstCards() {
+        launch {
+            sortedReferendaFlow
+                .filter { it.isNotEmpty() }
+                .first() // Await while list of cards will be not empty
+
+            loadContentForCardsFromPosition(0)
+        }
+    }
+
     private fun skipCard() {
         _skipCardEvent.sendEvent()
     }
 
-    private fun writeVote(position: Int) = launch {
-        cardsFlow.first()
-            .getOrNull(0)
-            ?.let { showMessage("Not implemented yet") }
+    private fun writeVote(position: Int, voteType: VoteType) {
+        if (isVotingInProgress.value) return
+        isVotingInProgress.value = true
+
+        launch {
+            val referendum = sortedReferendaFlow.value.getOrNull(position) ?: return@launch
+
+            if (!interactor.isVotingPowerAvailable()) {
+                val response = tinderGovVoteRequester.awaitResponse(TinderGovVoteRequester.Request(referendum.id.value))
+
+                if (!response.success) {
+                    _rewindCardEvent.sendEvent()
+                    isVotingInProgress.value = false
+                    return@launch
```

Comment:
Same here. If we leave "set voting power" screen, loading wont stop

Reply:
For this case loading will stop since I call `isVotingInProgress.value = false`